exchanges/bybit/bybit_websocket.go

```diff
@@ -0,0 +1,495 @@
+package bybit
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/common/crypto"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	exchange "github.com/thrasher-corp/gocryptotrader/exchanges"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/order"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/orderbook"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/stream"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/ticker"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/trade"
+	"github.com/thrasher-corp/gocryptotrader/log"
+)
+
+const (
+	bybitWSBaseURL      = "wss://stream.bybit.com/"
+	wsSpotPublicTopicV2 = "spot/quote/ws/v2"
+	wsSpotPrivate       = "spot/ws"
+	bybitWebsocketTimer = 30 * time.Second
+	wsOrderbook         = "depth"
+	wsTicker            = "bookTicker"
+	wsTrades            = "trade"
+	wsKlines            = "kline"
+
+	wsAccountInfoStr = "outboundAccountInfo"
+	wsOrderStr       = "executionReport"
+	wsOrderFilledStr = "ticketInfo"
+
+	sub    = "sub"    // event for subscribe
+	cancel = "cancel" // event for unsubscribe
+)
+
+var comms = make(chan stream.Response)
```

Comment:
There might be a way of making a custom struct type for this channel that also includes an asset type. 

```go
type resp struct {
	raw stream.Response
	asset asset.Item
}

var comms = make(chan resp)
```

Then when setting up a connection to each individual asset type we can call wsReadData with an extended parameter

```go
func (by *Bybit) wsReadData(ws stream.Connection, assetItem asset.Item) {
	by.Websocket.Wg.Add(1) // This should be called before go routine genration.
	defer by.Websocket.Wg.Done()
	for {
		respwow := ws.ReadMessage()
		if respwow.Raw == nil {
			return
		}
		comms <- resp{raw: respwow, asset: assetItem}
	}
}
```

Then we can consolidate ws_cfutures, ws_futures, ws_ufutures code into one by calling wsHandleData with that extended parameter set.

```go 
func (by *Bybit) wsHandleData(respRaw []byte, assetItem asset.Item) error {
	var multiStreamData map[string]interface{}
	err := json.Unmarshal(respRaw, &multiStreamData)
	if err != nil {
		return err
	}

	if method, ok := multiStreamData["event"].(string); ok {
		if strings.EqualFold(method, sub) {
			return nil
		}
		if strings.EqualFold(method, cancel) {
			return nil
		}
	}
```

Thus reducing a lot of the code and only using code from bybit_websocket.go

If this is a hard explanation I can explain it further or open a PR against your code if you need it.

Reply:
Hello @shazbert
I have understood what you are trying to say here. 
But I think we should keep such changes for the latter stage when we add full support for WS CMF, USDT, and Futures. 

Here are more reasons to do so.
1. wsHandleData is not exactly the same for WS Spot and other assets. Though we can do such optimization for WS CMF, USDT and Futures as they have most of the code same for wsHandleData.

2. WS support for CMF, USDT, and Futures is not 100% tested as @thrasher- suggested me to pause its development and keep the code related to WS CMF, USDT, and Futures in PR so that code can be used when we have support for multiple WS connections for different asset types in the core architecture of GCT.

I am open to discussion on it and please let me know your thoughts on it.

Reply:
Thanks for letting me know. I am happy to defer this and have this included in a separate PR.